File: phases/03-data-structures/04-deques-ring-buffers-circular-queues/code/main.c
/* main.c — deque, overwriting trace buffer, and lock-free SPSC ring. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdatomic.h>
#include <pthread.h>
#include <assert.h>
#include <time.h>

/* ============================================================ */
/* Deque using uncapped indices                                  */
/* ============================================================ */

typedef struct {
    int    *buf;
    size_t  head;       /* uncapped */
    size_t  tail;       /* uncapped */
    size_t  mask;
} Deque;

static void dq_init(Deque *d, size_t min_cap) {
    size_t cap = 1; while (cap < min_cap) cap *= 2;
    d->buf = malloc(cap * sizeof(int));
    d->head = d->tail = 0;
    d->mask = cap - 1;
}
static size_t dq_len(const Deque *d) { return d->tail - d->head; }
static int    dq_full(const Deque *d) { return dq_len(d) == d->mask + 1; }

static void dq_push_back(Deque *d, int x)  { assert(!dq_full(d)); d->buf[d->tail++ & d->mask] = x; }
static void dq_push_front(Deque *d, int x) { assert(!dq_full(d)); d->buf[--d->head & d->mask] = x; }
static int  dq_pop_back(Deque *d)          { assert(dq_len(d)); return d->buf[--d->tail & d->mask]; }
static int  dq_pop_front(Deque *d)         { assert(dq_len(d)); return d->buf[d->head++ & d->mask]; }
static void dq_free(Deque *d)              { free(d->buf); }

/* ============================================================ */
/* Overwriting trace buffer                                      */
/* ============================================================ */

typedef struct {
    int    *buf;
    size_t  head, tail, mask, len, capw;
} Trace;

static void tr_init(Trace *t, size_t cap) {
    size_t c = 1; while (c < cap) c *= 2;
    t->buf = malloc(c * sizeof(int));
    t->head = t->tail = t->len = 0;
    t->mask = c - 1; t->capw = c;
}
static void tr_push(Trace *t, int x) {
    t->buf[t->tail] = x;
    t->tail = (t->tail + 1) & t->mask;
    if (t->len == t->capw) t->head = (t->head + 1) & t->mask;
    else                   t->len++;
}
static void tr_dump(const Trace *t, const char *label) {
    printf("  %s [", label);
    for (size_t i = 0; i < t->len; ++i)
        printf("%d%s", t->buf[(t->head + i) & t->mask], i+1 < t->len ? ", " : "");
    printf("]\n");
}
static void tr_free(Trace *t) { free(t->buf); }

/* ============================================================ */
/* Lock-free SPSC ring buffer                                    */
/* ============================================================ */

#define SPSC_CAP 1024                       /* power of 2 */
typedef struct {
    int                 buf[SPSC_CAP];
    _Atomic size_t      head;               /* consumer-owned */
    _Atomic size_t      tail;               /* producer-owned */
} SPSC;

static int spsc_try_push(SPSC *q, int x) {
    size_t t = atomic_load_explicit(&q->tail, memory_order_relaxed);
    size_t h = atomic_load_explicit(&q->head, memory_order_acquire);
    if (t - h == SPSC_CAP) return 0;        /* full */
    q->buf[t & (SPSC_CAP - 1)] = x;
    atomic_store_explicit(&q->tail, t + 1, memory_order_release);
    return 1;
}

static int spsc_try_pop(SPSC *q, int *out) {
    size_t h = atomic_load_explicit(&q->head, memory_order_relaxed);
    size_t t = atomic_load_explicit(&q->tail, memory_order_acquire);
    if (h == t) return 0;                   /* empty */
    *out = q->buf[h & (SPSC_CAP - 1)];
    atomic_store_explicit(&q->head, h + 1, memory_order_release);
    return 1;
}

static SPSC g_queue;
static const size_t N_ITEMS = 1000000;

static void *producer(void *arg) {
    (void)arg;
    for (size_t i = 0; i < N_ITEMS; ) {
        if (spsc_try_push(&g_queue, (int)i)) i++;
    }
    return NULL;
}

static void *consumer(void *arg) {
    long *sum_out = (long *)arg;
    long sum = 0;
    for (size_t i = 0; i < N_ITEMS; ) {
        int x;
        if (spsc_try_pop(&g_queue, &x)) { sum += x; i++; }
    }
    *sum_out = sum;
    return NULL;
}

int main(void) {
    /* Deque demo */
    printf("== Deque ==\n");
    Deque d; dq_init(&d, 16);
    dq_push_back(&d, 1);
    dq_push_back(&d, 2);
    dq_push_back(&d, 3);
    dq_push_front(&d, 0);
    printf("  after pushes: len=%zu\n", dq_len(&d));
    printf("  pop_front -> %d (expect 0)\n", dq_pop_front(&d));
    printf("  pop_back  -> %d (expect 3)\n", dq_pop_back(&d));
    printf("  pop_front -> %d (expect 1)\n", dq_pop_front(&d));
    printf("  pop_front -> %d (expect 2)\n", dq_pop_front(&d));
    printf("  empty: len=%zu\n", dq_len(&d));
    dq_free(&d);

    /* Trace buffer demo */
    printf("\n== Overwriting trace buffer (cap=8) ==\n");
    Trace t; tr_init(&t, 8);
    for (int i = 1; i <= 5; ++i) tr_push(&t, i);
    tr_dump(&t, "after 1..5      :");
    for (int i = 6; i <= 12; ++i) tr_push(&t, i);
    tr_dump(&t, "after 1..12     :");
    tr_free(&t);

    /* SPSC demo */
    printf("\n== Lock-free SPSC ring buffer ==\n");
    pthread_t p, c;
    long sum = 0;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    pthread_create(&p, NULL, producer, NULL);
    pthread_create(&c, NULL, consumer, &sum);
    pthread_join(p, NULL);
    pthread_join(c, NULL);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    long expected = ((long)(N_ITEMS - 1) * N_ITEMS) / 2;
    printf("  consumed %zu items in %.3f s  (%.1f Mitems/s)\n",
           N_ITEMS, elapsed, N_ITEMS / 1e6 / elapsed);
    printf("  checksum %ld (expected %ld) — %s\n",
           sum, expected, sum == expected ? "OK" : "MISMATCH!");

    return 0;
}

File: phases/03-data-structures/04-deques-ring-buffers-circular-queues/code/main.rs
//! main.rs — Rust deque + lock-free SPSC ring buffer with explicit memory orderings.
//!
//! Build: `rustc -O main.rs && ./main`

use std::collections::VecDeque;
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;

/// Lock-free SPSC ring buffer for i32. Power-of-2 capacity.
pub struct SpscRing {
    buf: Box<[std::cell::UnsafeCell<i32>]>,
    head: AtomicUsize,
    tail: AtomicUsize,
    mask: usize,
}

unsafe impl Sync for SpscRing {}

impl SpscRing {
    pub fn new(cap: usize) -> Self {
        assert!(cap.is_power_of_two(), "cap must be a power of 2");
        let buf: Vec<std::cell::UnsafeCell<i32>> =
            (0..cap).map(|_| std::cell::UnsafeCell::new(0)).collect();
        SpscRing {
            buf: buf.into_boxed_slice(),
            head: AtomicUsize::new(0),
            tail: AtomicUsize::new(0),
            mask: cap - 1,
        }
    }

    /// Producer-side. Returns false if full.
    pub fn try_push(&self, x: i32) -> bool {
        let t = self.tail.load(Ordering::Relaxed);
        let h = self.head.load(Ordering::Acquire);
        if t.wrapping_sub(h) == self.mask + 1 { return false; }
        unsafe { *self.buf[t & self.mask].get() = x; }
        self.tail.store(t.wrapping_add(1), Ordering::Release);
        true
    }

    /// Consumer-side. Returns None if empty.
    pub fn try_pop(&self) -> Option<i32> {
        let h = self.head.load(Ordering::Relaxed);
        let t = self.tail.load(Ordering::Acquire);
        if h == t { return None; }
        let x = unsafe { *self.buf[h & self.mask].get() };
        self.head.store(h.wrapping_add(1), Ordering::Release);
        Some(x)
    }
}

fn main() {
    // VecDeque demo
    println!("== VecDeque (Rust's ring-buffer deque) ==");
    let mut d: VecDeque<i32> = VecDeque::new();
    d.push_back(1); d.push_back(2); d.push_back(3);
    d.push_front(0);
    println!("  after pushes: {:?}", d);
    println!("  pop_front -> {:?} (expect Some(0))", d.pop_front());
    println!("  pop_back  -> {:?} (expect Some(3))", d.pop_back());
    println!("  remaining: {:?}", d);

    // SPSC ring
    println!("\n== Lock-free SPSC ring buffer ==");
    const N: usize = 1_000_000;
    let q = Arc::new(SpscRing::new(1024));
    let qp = Arc::clone(&q);
    let qc = Arc::clone(&q);

    let t0 = std::time::Instant::now();
    let producer = thread::spawn(move || {
        let mut i: i32 = 0;
        while (i as usize) < N {
            if qp.try_push(i) { i += 1; }
        }
    });
    let consumer = thread::spawn(move || {
        let mut sum: i64 = 0;
        let mut count = 0;
        while count < N {
            if let Some(x) = qc.try_pop() {
                sum += x as i64;
                count += 1;
            }
        }
        sum
    });
    producer.join().unwrap();
    let sum = consumer.join().unwrap();
    let elapsed = t0.elapsed();

    let expected = ((N as i64 - 1) * N as i64) / 2;
    println!("  consumed {} items in {:.3?}  ({:.1} Mitems/s)",
             N, elapsed, N as f64 / 1e6 / elapsed.as_secs_f64());
    println!("  checksum {} (expected {}) — {}",
             sum, expected, if sum == expected { "OK" } else { "MISMATCH" });
}
